dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,475 @@
+
+logger = logging.getLogger(__name__)
+
+
+def _train(model, data, max_iter=1, dry_run=False, scorer=None,
+           s=1, i=1, k='_', **fit_kwargs):
+    """
+    Train function. Returns validation score.
+
+    Parameters
+    ----------
+    model : sklearn model to train
+    X, y, X_val, y_val : dask arrays that are training and validation sets
+    max_iter : float, number of times to call partial_fit
+    dry_run : whether to call partial_fit and score or not
+    s, i, k : int or string, used for logging
+    fit_kwargs : dict, kwargs to pass to partial_fit
+
+    Takes max_iters as parameters, which are treated as a "scarce resource" --
+    Hyperband is trying to minimize the total iters use for max_iters.
+
+    Returns
+    -------
+    val_score : float. Validation score
+    times : dict. keys of ``fit_time`` and ``score_time``, values in seconds.
+
+    """
+    for k, v in data.items():
+        for i, vi in enumerate(v):
+            if isinstance(vi, np.ndarray):
+                data[k][i] = da.from_array(vi, chunks=10)
+    X, y = data['train']
+    X_val, y_val = data['val']
+    start_time = default_timer()
+
+    for iter in range(int(max_iter)):
+        msg = ("Training model %d in bracket %d iteration %d. "
+               "This model is %0.2f trained for this bracket")
+        logger.info(msg, k, s, i, percent=iter * 100.0 / max_iter)
+        if not dry_run:
+            model.partial_fit(X, y, **fit_kwargs)
+    fit_time = default_timer() - start_time
+    msg = ("Training model %s for %d partial_fit calls took %0.2f seconds")
+    logger.info(msg, k, max_iter, fit_time)
+
+    start_time = default_timer()
+    score = scorer(model, X_val, y_val) if not dry_run else np.random.rand()
+    score_time = default_timer() - start_time
+    msg = "Fitting model {k} took {fit} seconds and scoring took {score} seconds"
+    logger.info(msg.format(k=k, fit=fit_time, score=score_time))
+    return score, {'score_time': score_time, 'fit_time': fit_time}
+
+
+def _top_k(choose_from, eval_with, k=1):
+    assert k > 0, "k must be positive"
+    assert set(choose_from.keys()) == set(eval_with.keys())
+
+    keys = list(choose_from.keys())
+    assert k <= len(keys)
+
+    eval_with = [eval_with[key] for key in keys]
+    idx = np.argsort(eval_with)
+
+    return {keys[i]: choose_from[keys[i]] for i in idx[-k:]}
+
+
+def _random_choice(params, n=1):
+    configs = ParameterSampler(params, n)
+    return configs
+
+
+def rec_type(d):
+    if isinstance(d, (float, int, str, np.ndarray, da.Array)):
+        return type(d)
+    if isinstance(d, list):
+        return [rec_type(di) for di in d]
+    if isinstance(d, dict):
+        return {k: rec_type(v) for k, v in d.items()}
+    return type(d)
+
+
+def _successive_halving(params, model, data, n=None, r=None, s=None,
+                        shared=None, eta=3, _prefix='', dry_run=False,
+                        scorer=None,
+                        **fit_kwargs):
```

Comment:
It would help future maintainers a bit if functions like these had docstrings describing what they do.